### PR TITLE
Implements md5()

### PR DESCRIPTION
**Pull request by ike709**
image

Since md5() can take a file or a text we first try to load the arg as a file converted to a string (largely copying file2text), then if that's null we just hash the arg string directly.

### CODE REVIEW

**Comment by ike709** [comment]
Turns out those dashes are because of BitConverter so give me a sec to replace it with an old-fashioned StringBuilder

File: DMCompiler/DMStandard/_Standard.dm [hunk 1]
```diff
@@ -72,6 +72,7 @@ proc/view(Dist = 4, Center = usr)
 proc/viewers(Depth, Center = usr)
 proc/walk(Ref, Dir, Lag = 0, Speed = 0)
 proc/walk_to(Ref, Trg, Min = 0, Lag = 0, Speed = 0)
+proc/md5(T)
```
Comment (wixoaGit) [inline DMCompiler/DMStandard/_Standard.dm:75]: Keep this in alphabetic order

File: OpenDreamServer/Dream/Procs/Native/DreamProcNative.cs [hunk 2]
```diff
@@ -77,6 +77,7 @@ static class DreamProcNative {
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_viewers);
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_walk);
             root.SetNativeProc(DreamProcNativeRoot.NativeProc_walk_to);
+            root.SetNativeProc(DreamProcNativeRoot.NativeProc_md5);
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNative.cs:80]: Same here

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 3]
```diff
@@ -1231,5 +1233,28 @@ public static DreamValue NativeProc_viewers(DreamObject instance, DreamObject us
 
             return DreamValue.Null;
         }
+        
+        [DreamProc("md5")]
+        [DreamProcParameter("T", Type = DreamValueType.String | DreamValueType.DreamResource)]
+        public static DreamValue NativeProc_md5(DreamObject instance, DreamObject usr, DreamProcArguments arguments) {
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs:1239]: And here

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 4]
```diff
@@ -1231,5 +1233,28 @@ public static DreamValue NativeProc_viewers(DreamObject instance, DreamObject us
...
+            if(arguments.ArgumentCount > 1) throw new Exception("md5() only takes one argument");
+            DreamValue arg = arguments.GetArgument(0, "T");
+            DreamResource resource;
+
+            if (arg.Type == DreamValueType.String) {
+                resource = Program.DreamResourceManager.LoadResource(arg.GetValueAsString());
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs]: Giving `md5()` a string always hashes the string, it doesn't attempt to load it as a file.

File: OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs [hunk 5]
```diff
@@ -1231,5 +1233,28 @@ public static DreamValue NativeProc_viewers(DreamObject instance, DreamObject us
...
+            } else {
+                resource = arg.GetValueAsDreamResource();
+            }
+
+            string text = resource.ReadAsString();
+            if (text == null) text = arg.GetValueAsString();
```
Comment (wixoaGit) [inline OpenDreamServer/Dream/Procs/Native/DreamProcNativeRoot.cs]: Lines 1242 to 1251 could be replaced with
```cs
string text;
if (arg.TryGetValueAsDreamResource(out DreamResource resource)) {
    text = resource.ReadAsString();
} else if (!arg.TryGetValueAsString(out text)) {
    return DreamValue.Null; //Return null if arg is neither a file or a string
}
```

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Comment by ike709** [comment]
@wixoaGit Addressed, the format of the output also matches byond exactly now.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**